data_collection/timeprof_matrix_bot.py

```diff
@@ -478,24 +460,45 @@ async def run_at(self, dt, coro):
         await self.wait_until(dt)
         return await coro
 
-    def create_next_sample_time(self, prev_sample_time, rate):
+    def draw_next_ping_time(self, prev_sample_time, rate):
         interval = np.random.exponential(scale=rate)
         next_sample_time = prev_sample_time + timedelta(minutes=interval)
         return next_sample_time
 
-    def schedule_next_sample(self, user_id, sample_time):
+    async def ping(self, user_id, sample_time):
+        loop = asyncio.get_event_loop()
+
+        if self.database.get_user_state(user_id) == STATE_ACTIVITY_WAIT:
+            loop.create_task(self.send_room_message("Previous sample unanswered, saving placeholder label...", room_id))
+            prev_sample_time = self.database.get_next_sample_time(user_id)
+            self.database.save_sample(user_id, prev_sample_time, "EMPTY")
+            self.database.set_user_state(user_id, STATE_NONE)
+
+        loop.create_task(self.prompt_user_activity(user_id))
+
+        rate = self.database.get_rate(user_id)
+        new_ping_time = self.draw_next_ping_time(sample_time, rate)
+        self.schedule_ping(self, user_id, new_ping_time)
+
+    def schedule_ping(self, user_id, ping_time):
         loop = asyncio.get_event_loop()
-        self.database.set_next_sample_time(user_id, sample_time)
-        loop.create_task(self.run_at(sample_time, self.collect_user_activity(user_id)))
-        logging.info("Scheduled new sample time at {}".format(sample_time))
+        logging.info("Scheduling next ping at {}".format(ping_time))
+        self.database.set_next_sample_time(user_id, ping_time)
```

Comment:
Let's add a function "set_prev_sample_time" and set it to the return value of get_next_sample_time before this line?